### ISSUE TITLE
Implement Stellar

### ISSUE THREAD

**Issue by tsusanka** [trezor/trezor-core#202]

**Comment by tsusanka** [comment]
@zulucrypto if it's okay I'd like to collect some questions regarding the trezor-mcu Stellar implementation here, because I'm writing the same for trezor-core and some things might be unclear.

**Comment by tsusanka** [comment]
@zulucrypto On the trezor-python side in the `stellar_sign_transaction` function you've decided to use provide the transaction details in base64 structure called envelope.

I've noticed this term in a various stellar documentation. Is that somewhat a standard in Stellar? Can some of the stellar apps create this on your behalf, obvisouly without having the keys?

**Comment by zulucrypto** [comment]
@tsusanka - Yes the "envelope" is what they call the entire transaction. Typically an app will create the transaction and send it to the Trezor to be signed.

There's an online tool for building transactions here: https://www.stellar.org/laboratory/#txbuilder

Documentation on the XDR format is a bit spotty. For the most part, I've been looking at the XDR files and generating test transactions.

All XDR template files: https://github.com/stellar/stellar-core/tree/master/src/xdr

`Stellar-transaction.x` will probably be the most useful: https://github.com/stellar/stellar-core/blob/master/src/xdr/Stellar-transaction.x

Let me know if there's anything I can do in terms of examples or more docs!

**Comment by prusnak** [comment]
JFYI I removed `Stellar{SignMessage,MessageSignature,VerifyMessage}` in https://github.com/trezor/trezor-mcu/commit/ae82735b6b7ecbe46a0baf3243ccd0f3e73ca21d and https://github.com/trezor/trezor-common/commit/fb662e53b129b250149d982259696531ff4c2b95, so let's not deal with these for now.

**Comment by tsusanka** [comment]
Also for reference: the sign/verify is implemented in the [tsusanka/stellar-signverify](https://github.com/trezor/trezor-core/tree/tsusanka/stellar-signverify) branch, but will not be part of the Stellar PR

**Comment by matejcik** [comment]
also for reference, python-trezor drops sign/verify in https://github.com/trezor/python-trezor/commit/4b7926268e7f508595bf3fa43456dd3254183056 and https://github.com/trezor/python-trezor/commit/7e83eb11e51a63ab31a4c3604c9ad630339de322

**Comment by tsusanka** [comment]
@zulucrypto out of curiosity where did you take the test data in python-trezor? Did you get them from some other repository or did you just assert what trezor returned?

e.g. https://github.com/trezor/python-trezor/blob/master/trezorlib/tests/device_tests/test_msg_stellar_sign_transaction.py#L40 

(btw I've also contacted you on gitter, but I can copy here, in case you're not using that)

**Comment by zulucrypto** [comment]
@tsusanka - I'm also working on a PHP SDK for Stellar and have a collection of unit and functional tests so I copied the signature generated by those tests.

Let me know if they would be helpful to you and I can write up some instructions to get you going!

**Comment by tsusanka** [comment]
@zulucrypto Ah, great. Well, I'd like to add a few more tests, the ideal state is to have at least one test for each operation. I can have a look what the trezor-mcu returns now and assume that is correct (which most likely is) and simply assert that signature (that is still beneficial for future modifications and mainly for trezor-core). But obviously, it would be better to have test data from some other party.

So tl;dr: would you like to add some more tests based on your sdk data?

**Comment by zulucrypto** [comment]
@tsusanka - Sure thing, I'll work on adding some more tests to python-trezor and open a PR for them

**Comment by tsusanka** [comment]
That's awesome. Let me know if you need anything

**Comment by tsusanka** [comment]
@zulucrypto I was looking at the list of stellar [operations](https://www.stellar.org/developers/guides/concepts/list-of-operations.html) and I've noticed two differences:
1) `BumpSequence` is not in the list. That is because it is quite new operation (https://github.com/stellar/stellar-core/pull/1533) so I guess it was not documented yet?
2) `Inflation` is not in the mcu implementation. Any reason for that?

**Comment by tsusanka** [comment]
Oh, I see, the inflation operation is used once a week to get the inflation payout, so it's mainly used by the inflation pools and doesn't make much sense to implement it in trezor. Is that about right?

**Comment by zulucrypto** [comment]
@tsusanka 

`BumpSequence` - Yes, it was under discussion when I was working on the code, but the XDR format seemed stable so I included it in the PR.

`Inflation` - Yes, any user can submit this operation to the network to trigger the inflation calculation, so there wouldn't be any point in using a Trezor to secure it. :)
